argocd cli segfault if ignoreDifferences is set

If you omit the `Group:` in an ignoreDifferences setting, the argocd cli will segfault. 

```
  ignoreDifferences:
  - kind: Secret
    jsonPointers:
    - /data/token
```

```
$ argocd app diff control-vault-crd
panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 1 [running]:
github.com/argoproj/argo-cd/util/diff.NormalizeSecret(0xc000158210)
	/go/src/github.com/argoproj/argo-cd/util/diff/diff.go:308 +0x637
github.com/argoproj/argo-cd/util/diff.Normalize(0xc000158210, 0x2719680, 0xc00063c4c0)
	/go/src/github.com/argoproj/argo-cd/util/diff/diff.go:266 +0x1b7
github.com/argoproj/argo-cd/util/diff.Diff(0xc000158210, 0xc0003d22a0, 0x2719680, 0xc00063c4c0, 0x2719680)
	/go/src/github.com/argoproj/argo-cd/util/diff/diff.go:40 +0x1c6
github.com/argoproj/argo-cd/cmd/argocd/commands.NewApplicationDiffCommand.func1(0xc000016900, 0xc000453520, 0x1, 0x1)
	/go/src/github.com/argoproj/argo-cd/cmd/argocd/commands/app.go:759 +0xadc
github.com/spf13/cobra.(*Command).execute(0xc000016900, 0xc0004534e0, 0x1, 0x1, 0xc000016900, 0xc0004534e0)
	/go/src/github.com/spf13/cobra/command.go:702 +0x2d3
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001efd40, 0xc0004ec000, 0xc0000b7b30, 0xc0003e5a40)
	/go/src/github.com/spf13/cobra/command.go:783 +0x2dc
github.com/spf13/cobra.(*Command).Execute(0xc0001efd40, 0xc000539f88, 0xc0000460b8)
	/go/src/github.com/spf13/cobra/command.go:736 +0x2b
main.main()
	/go/src/github.com/argoproj/argo-cd/cmd/argocd/main.go:14 +0x27
```

```
argocd: v0.12.0+cd4bb25
  BuildDate: 2019-03-22T19:59:10Z
  GitCommit: cd4bb2553d4d3b9a8ba808adbd65c15ce11d9cfa
  GitTreeState: clean
  GoVersion: go1.11.4
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v0.12.0+cd4bb25
  BuildDate: 2019-03-22T19:58:47Z
  GitCommit: cd4bb2553d4d3b9a8ba808adbd65c15ce11d9cfa
  GitTreeState: clean
  GoVersion: go1.11.4
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: 0.13.1
```

--- a/test/e2e/app_management_test.go
+++ b/test/e2e/app_management_test.go
@@ -65,6 +65,28 @@ func getTestApp() *v1alpha1.Application {
 	}
 }
 
+func createAndSync(t *testing.T, appPath string) *v1alpha1.Application {
+	app := getTestApp()
+	app.Spec.Source.Path = appPath
+
+	app, err := fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Create(app)
+	assert.NoError(t, err)
+
+	_, err = fixture.RunCli("app", "sync", app.Name)
+	assert.NoError(t, err)
+
+	_, err = fixture.RunCli("app", "wait", app.Name, "--sync", "--timeout", "5")
+	assert.NoError(t, err)
+
+	app, err = fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Get(app.Name, metav1.GetOptions{})
+	assert.NoError(t, err)
+	return app
+}
+
+func createAndSyncDefault(t *testing.T) *v1alpha1.Application {
+	return createAndSync(t, guestbookPath)
+}
+
 func TestAppCreation(t *testing.T) {
 	fixture.EnsureCleanState()
 
@@ -123,18 +145,9 @@ func TestAppDeletion(t *testing.T) {
 func TestTrackAppStateAndSyncApp(t *testing.T) {
 	fixture.EnsureCleanState()
 
-	app, err := fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Create(getTestApp())
-	assert.NoError(t, err)
+	app := createAndSyncDefault(t)
 
-	// sync app and make sure it reaches InSync state
-	_, err = fixture.RunCli("app", "sync", app.Name)
-	assert.NoError(t, err)
 	assertAppHasEvent(t, app, "sync", argo.EventReasonResourceUpdated)
-
-	WaitUntil(t, func() (done bool, err error) {
-		app, err = fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Get(app.ObjectMeta.Name, metav1.GetOptions{})
-		return err == nil && app.Status.Sync.Status == v1alpha1.SyncStatusCodeSynced, err
-	})
 	assert.Equal(t, v1alpha1.SyncStatusCodeSynced, app.Status.Sync.Status)
 	assert.True(t, app.Status.OperationState.SyncResult != nil)
 	assert.True(t, app.Status.OperationState.Phase == v1alpha1.OperationSucceeded)
@@ -213,21 +226,14 @@ func TestComparisonFailsIfClusterNotAdded(t *testing.T) {
 func TestArgoCDWaitEnsureAppIsNotCrashing(t *testing.T) {
 	fixture.EnsureCleanState()
 
-	app := getTestApp()
-
-	// deploy app and make sure it is healthy
-	app, err := fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Create(app)
-	assert.NoError(t, err)
-
-	_, err = fixture.RunCli("app", "sync", app.Name)
-	assert.NoError(t, err)
+	app := createAndSyncDefault(t)
 
 	WaitUntil(t, func() (done bool, err error) {
 		app, err = fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Get(app.ObjectMeta.Name, metav1.GetOptions{})
 		return err == nil && app.Status.Sync.Status == v1alpha1.SyncStatusCodeSynced && app.Status.Health.Status == v1alpha1.HealthStatusHealthy, err
 	})
 
-	_, err = fixture.RunCli("app", "set", app.Name, "--path", "crashing-guestbook")
+	_, err := fixture.RunCli("app", "set", app.Name, "--path", "crashing-guestbook")
 	assert.NoError(t, err)
 
 	_, err = fixture.RunCli("app", "sync", app.Name)
@@ -242,17 +248,7 @@ func TestArgoCDWaitEnsureAppIsNotCrashing(t *testing.T) {
 func TestManipulateApplicationResources(t *testing.T) {
 	fixture.EnsureCleanState()
 
-	app := getTestApp()
-
-	// deploy app and make sure it is healthy
-	app, err := fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Create(app)
-	assert.NoError(t, err)
-
-	_, err = fixture.RunCli("app", "sync", app.Name)
-	assert.NoError(t, err)
-
-	_, err = fixture.RunCli("app", "wait", app.Name)
-	assert.NoError(t, err)
+	app := createAndSyncDefault(t)
 
 	manifests, err := fixture.RunCli("app", "manifests", app.Name, "--source", "live")
 	assert.NoError(t, err)
@@ -286,6 +282,30 @@ func TestManipulateApplicationResources(t *testing.T) {
 		app, err = fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Get(app.ObjectMeta.Name, metav1.GetOptions{})
 		return err == nil && app.Status.Sync.Status == v1alpha1.SyncStatusCodeOutOfSync, err
 	})
+}
+
+func TestAppWithSecrets(t *testing.T) {
+	fixture.EnsureCleanState()
+
+	app := createAndSync(t, "secrets")
+
+	app.Spec.IgnoreDifferences = []v1alpha1.ResourceIgnoreDifferences{{
+		Kind: kube.SecretKind, JSONPointers: []string{"/data/username"},
+	}}
+
+	closer, client, err := fixture.ArgoCDClientset.NewApplicationClient()
+	assert.NoError(t, err)
+	defer util.Close(closer)
+	_, err = client.UpdateSpec(context.Background(), &application.ApplicationUpdateSpecRequest{
+		Name: &app.Name,
+		Spec: app.Spec,
+	})
+
+	assert.NoError(t, err)
+
+	diffOutput, err := fixture.RunCli("app", "diff", app.Name)
+	assert.NoError(t, err)
+	assert.Empty(t, diffOutput)
 }
 
 func TestResourceDiffing(t *testing.T) {
@@ -354,13 +374,8 @@ func TestEdgeCasesApplicationResources(t *testing.T) {
 	for name, appPath := range apps {
 		t.Run(fmt.Sprintf("Test%s", name), func(t *testing.T) {
 			fixture.EnsureCleanState()
-			app := getTestApp()
-			app.Spec.Source.Path = appPath
-			app, err := fixture.AppClientset.ArgoprojV1alpha1().Applications(fixture.ArgoCDNamespace).Create(app)
-			assert.NoError(t, err)
 
-			_, err = fixture.RunCli("app", "sync", app.Name)
-			assert.NoError(t, err)
+			app := createAndSync(t, appPath)
 
 			closer, client, err := fixture.ArgoCDClientset.NewApplicationClient()
 			assert.NoError(t, err)

--- a/test/e2e/testdata/secrets/secrets.yaml
+++ b/test/e2e/testdata/secrets/secrets.yaml
@@ -1,3 +1,4 @@
+---
 apiVersion: v1
 kind: Secret
 metadata:
@@ -6,3 +7,11 @@ stringData:
   username: test-username
 data:
   password: dGVzdC1wYXNzd29yZA==
+
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: test-secret2
+stringData:
+  username: test-username

--- a/util/argo/diff_normalizer.go
+++ b/util/argo/diff_normalizer.go
@@ -80,7 +80,7 @@ func NewDiffNormalizer(ignore []v1alpha1.ResourceIgnoreDifferences, overrides ma
 }
 
 // Normalize removes fields from supplied resource using json paths from matching items of specified resources ignored differences list
-func (n *normalizer) Normalize(un *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+func (n *normalizer) Normalize(un *unstructured.Unstructured) error {
 	matched := make([]normalizerPatch, 0)
 	for _, patch := range n.patches {
 		groupKind := un.GroupVersionKind().GroupKind()
@@ -92,12 +92,12 @@ func (n *normalizer) Normalize(un *unstructured.Unstructured) (*unstructured.Uns
 		}
 	}
 	if len(matched) == 0 {
-		return un, nil
+		return nil
 	}
 
 	docData, err := json.Marshal(un)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	for _, patch := range matched {
@@ -106,14 +106,14 @@ func (n *normalizer) Normalize(un *unstructured.Unstructured) (*unstructured.Uns
 			if err.Error() == noMatchingPathError {
 				continue
 			}
-			return nil, err
+			return err
 		}
 		docData = patchedData
 	}
 
 	err = json.Unmarshal(docData, un)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return un, nil
+	return nil
 }

--- a/util/argo/diff_normalizer_test.go
+++ b/util/argo/diff_normalizer_test.go
@@ -26,7 +26,7 @@ func TestNormalizeObjectWithMatchedGroupKind(t *testing.T) {
 	assert.Nil(t, err)
 	assert.True(t, has)
 
-	deployment, err = normalizer.Normalize(deployment)
+	err = normalizer.Normalize(deployment)
 	assert.Nil(t, err)
 	_, has, err = unstructured.NestedSlice(deployment.Object, "spec", "template", "spec", "containers")
 	assert.Nil(t, err)
@@ -44,10 +44,12 @@ func TestNormalizeNoMatchedGroupKinds(t *testing.T) {
 
 	deployment := kube.MustToUnstructured(test.DemoDeployment())
 
-	normalized, err := normalizer.Normalize(deployment)
+	err = normalizer.Normalize(deployment)
 	assert.Nil(t, err)
 
-	assert.Equal(t, deployment.Object, normalized.Object)
+	_, hasSpec, err := unstructured.NestedMap(deployment.Object, "spec")
+	assert.Nil(t, err)
+	assert.True(t, hasSpec)
 }
 
 func TestNormalizeMatchedResourceOverrides(t *testing.T) {
@@ -65,7 +67,7 @@ func TestNormalizeMatchedResourceOverrides(t *testing.T) {
 	assert.Nil(t, err)
 	assert.True(t, has)
 
-	deployment, err = normalizer.Normalize(deployment)
+	err = normalizer.Normalize(deployment)
 	assert.Nil(t, err)
 	_, has, err = unstructured.NestedSlice(deployment.Object, "spec", "template", "spec", "containers")
 	assert.Nil(t, err)

--- a/util/diff/diff.go
+++ b/util/diff/diff.go
@@ -29,7 +29,7 @@ type DiffResultList struct {
 }
 
 type Normalizer interface {
-	Normalize(un *unstructured.Unstructured) (*unstructured.Unstructured, error)
+	Normalize(un *unstructured.Unstructured) error
 }
 
 // Diff performs a diff on two unstructured objects. If the live object happens to have a
@@ -249,15 +249,6 @@ func Normalize(un *unstructured.Unstructured, normalizer Normalizer) {
 		return
 	}
 
-	if normalizer != nil {
-		normalized, err := normalizer.Normalize(un)
-		if err == nil {
-			un = normalized
-		} else {
-			log.Warnf("Failed to normalize %s/%s/%s: %v", un.GroupVersionKind(), un.GetNamespace(), un.GetName(), err)
-		}
-	}
-
 	// creationTimestamp is sometimes set to null in the config when exported (e.g. SealedSecrets)
 	// Removing the field allows a cleaner diff.
 	unstructured.RemoveNestedField(un.Object, "metadata", "creationTimestamp")
@@ -266,6 +257,13 @@ func Normalize(un *unstructured.Unstructured, normalizer Normalizer) {
 		NormalizeSecret(un)
 	} else if gvk.Group == "rbac.authorization.k8s.io" && (gvk.Kind == "ClusterRole" || gvk.Kind == "Role") {
 		normalizeRole(un)
+	}
+
+	if normalizer != nil {
+		err := normalizer.Normalize(un)
+		if err != nil {
+			log.Warnf("Failed to normalize %s/%s/%s: %v", un.GroupVersionKind(), un.GetNamespace(), un.GetName(), err)
+		}
 	}
 }
 
@@ -305,7 +303,7 @@ func NormalizeSecret(un *unstructured.Unstructured) {
 		return
 	}
 	if secret.Data != nil {
-		err = unstructured.SetNestedMap(un.Object, newObj["data"].(map[string]interface{}), "data")
+		err = unstructured.SetNestedField(un.Object, newObj["data"], "data")
 		if err != nil {
 			log.Warnf("failed to set secret.data: %v", err)
 			return

